### ISSUE TITLE
Object Keys: Support for arbitrary string as key.  

### ISSUE THREAD

**Issue by bedantaguru** [richfitz/storr#75]
During the time I was using `storr_rds` , I realised that key are stored as filename (at least in windows). Accordingly I think it was the root cause of  #70 . I have recently started migration of my code base form my own package (https://github.com/MadeInR/store) to storr. But I realised storr does not support yet arbitrary keys (I was using URLs) . I temporarily fixed it by constructing hash map of strings (URLs in my case). It would be nice if it is generalised to more broader approach where any arbitrary R object also can be treated as keys. I guess this may be easily be done through hash map.

**Comment by bedantaguru** [comment]
I have gone roughly through the code and realised that `mangle_key = T` will do the trick of using any arbitrary string as key. Also it will fix the case insensitivity problem (as mentioned in #70) in keys. Object as key is rather complicated demand. However mangle key is making the file path way bigger if key is large (example a long URL).

**Comment by richfitz** [comment]
Yes, that is all working as expected.  There is no function that can be used to convert an arbitrary string to a unique string in a reversible way.  We can hash (constant size, but irreversible) or encode (which will increase the size ~30%).

If you want to store objects against a key of a size that exceeds what you want to store on disk you should create a map in another namespace.

```
path <- tempfile()
st <- storr::storr_rds(path, mangle_key = TRUE)

url <- "https://longurlmaker.com/go?id=ganglingEasyURL93rangydrunning060121lastingDwarfurl61141URLcutStartURL1high7high194continuedfarZoffYATUClofty1lasting106URLHawk84Shorl3lofty2XZset6lengthy977f0ShrinkURL401Ne15lengthened15URLviURLcutlengthy60u0c36701stretchedFwdURL1greatspreadZoutod2e4EasyURL0farZreachingShrinkURL3stretchedrangyTinyLinkUrlTeaebelongate0ganglingbd11oShortURL168Smallrafarawayb24spunZout144ShrinkURL41Shim6noutstretcheda2frunningf4YepIt11TinyURL0running6301URL5da4d7great7stretchedb20gURlZielengthy102yEzURLzSitelutions00w72continued7fhighStartURL1protractedr180Shrinkrrangy71l8toweringXilShimShrinkr0Shim007804dulengthened0DigBigUrlTeadeepelongated7qTightURL9w51x0301URLc8ra62DoiopaURLZcoZuk3201f0Fly2lingering0ganglingyXil925sustainedShrinkURLu5jt01lofty11stringyUrlTeab00drawnZouttowering8DecentURLetallstretchingdistant907URLHawkenduring045647NanoRefsstretchlasting2n0EasyURLrangy61EzURLstretchinghdistantShrtnd6osspreadZoutNotLongelongatedc0w301URL301URL920Sitelutions02stringy6FhURLShortURL7ShortURLystretching311f1516051stretcheduTinyURL3cfarZoffv011NutshellURLdeeptall2protractedv0URLHawkelongate33Minilien56bnG8Lqf11NotLong1n707greatloftyrangy850EzURLNanoRef0v1827YATUC1rangy1301URL4continued6dm1URLPie1toweringea4enduringNanoRef744YATUC41URLZcoZukMyURL34farZoff0farZreachingw02TinyURL8m301URLx14r80enlargedd4lengthySnipURL50sustained835loftyenlarged1extensiveySHurlt6URLenduring78outstretched1TightURL5012WapURLflofty70deeplengthenedUrlTea0sustained0yShortURL2high375611spreadZoutdBeamZto1t4salasting6URL1b4a1301URL33Ulimitenduring13bShortURLvRedirx0f511d8801911k52Smallr710PiURLYepIt771303lankystretchingsastringy7cj4URLZcoZukRedirxgangling1063Shim0TinyLink9enlarged5running137Doiop0bShortlinkse26stringy00232stringyPiURL0tall71848stretchingDoiop6Sitelutions801160lengthylengthened1d1XilSitelutions7URLZcoZuk6rangyG8L1ShortURLf0101ShrinkURLlengthened2m7deep40NutshellURLa5lingeringq9lShrtndaSHurlcprotractedcsustained5dlsIsZgd114115lingering70r4119elongated1remoteURLCutterspunZout3enlargedURLCutterURLCutterv7q1p009hstretched1d05Shrtndx483e497ulLiteURLr3g"

short_marker <- "storr_github_repo"

st$set(short_marker, url, namespace = "url")
st$set(short_marker, TRUE, namespace = "visited")
```

**Comment by bedantaguru** [comment]
Thanks a lot @richfitz I finally got a clue and created a prototype package [resumable ](https://github.com/MadeInR/resumable).

The solution was rather simple as you pointed out. In you comment I replaced `short_marker` with hash of key object.  
```
st$set <- function(key, value, namespace) {
    if (missing(namespace))
      namespace <- stv$default_namespace
    key_h <- stv$hash_object(key)
    stk$set(key_h, key, namespace)
    stv$set(key_h, value, namespace)
  }
```
Thanks again for showing the path and this wonderful package. Maybe in near future this will be part of your package itself.
